hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/customfees/AdjustmentUtils.java

```diff
@@ -0,0 +1,232 @@
+        final var credits = new HashMap<AccountID, Pair<Long, TokenID>>();
+        for (final var entry : tokenChanges.entrySet()) {
+            final var account = entry.getKey();
+            final var amount = entry.getValue();
+            if (amount > 0 && account.equals(beneficiary)) {
+                credits.put(account, Pair.of(amount, tokenId));
+            }
+        }
+        for (final var entry : result.getInputHbarAdjustments().entrySet()) {
+            final var account = entry.getKey();
+            final var amount = entry.getValue();
+            if (amount > 0 && account.equals(beneficiary)) {
+                credits.put(account, Pair.of(amount, null));
+            }
+        }
+        return credits;
+    }
+
+    /**
+     * Adjusts hbar fees. It makes 2 adjustments a debit for sender and credit to collector.
+     * If there is already an entry merges the new change with it, otherwise creates a new entry.
+     * @param result The {@link AssessmentResult} object
+     * @param sender The sender account
+     * @param hbarFee The {@link CustomFee} object
+     */
+    public static void adjustHbarFees(final AssessmentResult result, final AccountID sender, final CustomFee hbarFee) {
+        final var hbarAdjustments = result.getHbarAdjustments();
+        final var collector = hbarFee.feeCollectorAccountId();
+        final var fixedSpec = hbarFee.fixedFee();
+        final var amount = fixedSpec.amount();
+        hbarAdjustments.merge(sender, -amount, Long::sum);
+        hbarAdjustments.merge(collector, amount, Long::sum);
+    }
+
+    /**
+     * Checks if the adjustment will trigger a custom fee.
+     * Custom fee is triggered if the fee is not self-denominated and the transfer is not a hbar transfer
+     * and the adjustment is a debit.
+     *
+     * @param denominatingTokenID     the token that is being used as denomination to pay the fee
+     * @param exemptDebits            the set of tokens that are exempt from custom fee charging, due to being self
+     *                                denominated in same level of custom fee assessment
+     * @return true if the adjustment will trigger a custom fee. False otherwise.
+     */
+    public static boolean couldTriggerCustomFees(
+            @NonNull final TokenID denominatingTokenID, @NonNull final Set<TokenID> exemptDebits) {
+        return !isExemptFromCustomFees(denominatingTokenID, exemptDebits);
+    }
+
+    /**
+     * Custom fee that is self-denominated is exempt from further custom fee charging.
+     *
+     * @param denominatingTokenID the token that is being used as denomination to pay the fee
+     * @param exemptDebits       the set of tokens that are exempt from custom fee charging, due to being self
+     *                           denominated in the same level of custom fee assessment
+     * @return true if the custom fee is self-denominated
+     */
+    private static boolean isExemptFromCustomFees(
```

Comment:
I haven't thought deeply, but if you are now merging all self-denominated (exempt) fixed fees into the current `op`, do we still need this?

Reply:
Don't we need this to check exemptions in same level ?

Reply:
As discussed removed